fem/tmop.hpp

```diff
@@ -726,20 +736,57 @@ class DiscreteAdaptTC : public TargetConstructor
    // Owned.
    AdaptivityEvaluator *adapt_eval;
 
+   void SetSerialDiscreteTargetBase(const GridFunction &tspec_);
+#ifdef MFEM_USE_MPI
+   void SetParDiscreteTargetBase(const ParGridFunction &tspec_);
+#endif
+   void SetTspecAtIndex(int idx, const GridFunction &tspec_);
+
 public:
    DiscreteAdaptTC(TargetType ttype)
       : TargetConstructor(ttype),
+        ncomp(0),
+        sizeidx(-1), skewidx(-1), aspectratioidx(-1), orientationidx(-1),
         tspec(), tspec_sav(), tspec_pert1h(), tspec_pert2h(), tspec_pertmix(),
-        tspec_fes(NULL),
+        ptspec_fes(NULL), tspec_fes(NULL), tspec_fesv(NULL),
         good_tspec(false), good_tspec_grad(false), good_tspec_hess(false),
         adapt_eval(NULL) { }
 
-   virtual ~DiscreteAdaptTC() { delete adapt_eval; }
+   virtual ~DiscreteAdaptTC()
+   {
+      delete adapt_eval;
+      delete tspec_fes;
+      delete tspec_fesv;
+   }
 
-   virtual void SetSerialDiscreteTargetSpec(GridFunction &tspec_);
+   /** @name Target specification methods.
+       The following methods are used to specify geometric parameters of the
+       targets when these parameters are given by discrete FE functions.
+       Note that every GridFunction given to the Set methods must use a
+       H1_FECollection of the same order. The number of components must
+       correspond to the type of geometric parameter and dimension.
+       Once the calls to the Set methods are complete, users are expected to
+       call a Finalize method.
+
+       @param[in] tspec_  Input values of a geometric parameter. Note that
+                          the methods in this class support only functions that
+                          use H1_FECollection collection of the same order. */
+   ///@{
+   virtual void SetSerialDiscreteTargetSpec(const GridFunction &tspec_);
+   virtual void SetSerialDiscreteTargetSize(const GridFunction &tspec_);
+   virtual void SetSerialDiscreteTargetSkew(const GridFunction &tspec_);
+   virtual void SetSerialDiscreteTargetAspectRatio(const GridFunction &tspec_);
```

Comment:
Could you add a short comment how many components are expected for a given dimension, and what the components represent in the skew / aspect ratio / orientation cases.

Reply:
Maybe is should be done in the general block comment above.